Use workload sets as per user request

Forgind opened the pull request:
This PR should enable installing and updating via workload sets. In particular, if you have mode set to workload sets (and don't use a rollback file), it will now look for the latest workload set (or latest stable if you don't specify to use previews) and install/update to that rather than updating manifests individually.

It now also supports `dotnet workload update --version x` where it will update the workload set version x.

Finally, it now will display the workload set it is using if it's set to use workload sets in the install state file when the user runs `dotnet workload list`.

All functionality was manually tested at some point, and there's a new unit test.

Forgind committed:
Support accessing install mode

Causes the dotnet workload update/install/repair commands to use a new GetWorkloadInstallMode method to decide whether to use workload sets or not. If they decide to use workload sets, they calculate manifest updates by reading from a dictionary that is not implemented.

Forgind commented on src/Cli/dotnet/commands/dotnet-workload/install/FileBasedInstaller.cs:
```diff
@@ -702,5 +702,12 @@ private bool PackHasInstallRecords(PackInfo packInfo)
+
+        public bool GetWorkloadInstallMode(SdkFeatureBand sdkFeatureBand)
+        {
+            string path = Path.Combine(WorkloadInstallType.GetInstallStateFolder(sdkFeatureBand, _dotnetDir), "default.json");
```
Probably don't need to abstract through IInstaller because the WorkloadInstallType.GetInstallStateFolder call is the same for both types (as is the rest of the code)

Forgind committed:
Simplify getting install mode

Forgind committed:
Update to use workload sets

Forgind committed:
Update to properly install via workload set

Forgind committed:
Undo List changes

Forgind committed:
Get real version

Forgind committed:
Implement workload update --version

Forgind committed:
Support rolling back workload set installation

Forgind committed:
Don't roll back for file-based

Forgind committed:
Small touchup

Forgind committed:
First stab at MSI-based install

joeloff commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs:
```diff
@@ -555,7 +634,14 @@ public void Shutdown()
 
         public PackageId GetManifestPackageId(ManifestId manifestId, SdkFeatureBand featureBand)
```
This is misleading, I'd suggest having two separate methods for manifest and workload set package, even though there similar - it will make it much easier to understand the code.

Forgind replied:
I'll admit you're right that it's a little misleading. I'm trying to balance making things clearer vs. reusing more code—since this method is called in the middle of where I need it, I'd have to either put a random if/else in the middle of that method (that is, move this up one stack frame) or copy most of the method into a separate method just for workload sets. I'd rather not lose any updates that come in the future, which is why I architected it this way. If you feel strongly, though, I can change that.

Forgind replied:
In GetManifestPackageDownloadsAsync and UpdateManifestWithVersionAsync, we shouldn't know whether it's a workload set manifest ID or something else.

Forgind committed:
Reimplement rolling back

Plus more correct MSI-based workload set installation

Forgind committed:
Little fixups

Forgind committed:
Overwrite workload set if present

Forgind committed:
Fix garbage collecting workload sets

Forgind committed:
Find workload set before installing it

Forgind committed:
Avoid elevation if we will noop in MSI-based operations to adjust the install state (#38117)

Forgind committed:
Support workload sets in list

Forgind committed:
Partial progress

Forgind committed:
Finish adding to install state

Forgind committed:
Remove InstallStateReader

Forgind committed:
Small fixes

Forgind committed:
More small fixes

Forgind committed:
Fix rollback behavior

Forgind committed:
Update some tests

Forgind committed:
Add test

Forgind committed:
Simplify workload set  version finding in list

joeloff commented on src/Cli/Microsoft.DotNet.Cli.Utils/PathUtility.cs:
```diff
@@ -97,6 +97,32 @@ public static bool TryDeleteDirectory(string directoryPath)
+                    dir = Path.GetDirectoryName(dir);
+                }
+            }
+            catch (Exception) { }
```
Is there a specific exception you get for non-empty directories. If permissions are invalid, do we want to ignore those or log them to make the user aware of the underlying problem?

Forgind replied:
I think both of those reasons are very possible. We should definitely always ignore an error due to a non-empty directory (that's an IOException). The other exception we're reasonably likely to hit (UnathorizedAccessException) I think we should ignore but mostly because this is a generic PathUtility method rather than a specific I-expect-to-be-able-to-delete-it method. If you want to make it just catch IOException, I'd probably have to handle UnauthorizedAccessException separately, but I don't mind doing that. Let me know what you think.

joeloff replied:
You could look at using `Directory.EnumerateFileSystemEntries().Any()` - it's lazy evaluation so you shouldn't incur a huge cost (unlike `GetFileSystemEntries`). That way, any other exception, including `IOException` will likely be a valid error we'd want the calling method to report.

Forgind replied:
I wanted to avoid doing too many expensive file system operations, but it sounds like Directory.EnumerateFileSystemEntries() is fast, so I'll use that and get rid of the try/catch. Thanks!

joeloff commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs:
```diff
@@ -182,6 +185,67 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
+
+        public void RollBackWorkloadSetInstallation()
+        {
+            if (_newWorkloadSetPath is not null)
```
This should probably check for ! nullorwhitespace since an empty string wouldn't make any sense here I think. Also, seems odd to record an internal state in a private member called `_newWorkloadSetPath`. It's only new at the time of being set, but then it's the actual value. Can we parameterize it at all?

Forgind replied:
I'd argue that if it's white space (and not null), we have a bug anyway, and we may as well try to go forward and crash so we at least find out afterwards.

What do you mean by parameterizing it? Like make it RollbackWorkloadSetInstallation(string) then pass the workload set path to that?

joeloff replied:
yup, add a string parameter is what I meant.

joeloff commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs:
```diff
@@ -182,6 +185,67 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
+
+            InstallAction plannedAction = PlanPackage(msi, state, installAction, installedVersion, out IEnumerable<string> relatedProducts);
+
+            if (plannedAction != InstallAction.None)
```
If there is no planned action, you don't need to execute the package. You also have to elevate before you can updat the dependent. I'd move the execute package and update call into the IF block

Forgind replied:
ExecutePackage basically just exits if plannedAction is InstallAction.None, and UpdateDependent Elevates separately, so it doesn't actually affect anything in this case, but I have no problem doing that.

joeloff commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs:
```diff
@@ -182,6 +185,67 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
+            VerifyPackage(msi);
+            DetectState state = DetectPackage(msi.ProductCode, out Version installedVersion);
+
+            if (installAction == InstallAction.Install)
```
This can be a removal as well right? I'd suggest renaming the parameter to `requestedAction`. PlanPackage can alter the requested action (or not). So after planning, if the planned action is still `InstallAction.Install` you'd be able to do this check. The code will work this way, but reading through it, it seems to be doing steps out of order

Forgind committed:
Couple comments

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/IWorkloadManifestUpdater.cs:
```diff
@@ -21,6 +21,8 @@ internal interface IWorkloadManifestUpdater
 
         void DeleteUpdatableWorkloadsFile();
+
+        void DownloadWorkloadSet(string version, DirectoryPath? offlineCache);
```
The workload manifest updater doesn't need to download a specific version of a workload set.  Its purpose is to check for the latest version available.  The installer implementations download (via NuGetPackageDownloader) the actual package to be installed (which will differ depending on whether the installer is the File-based or MSI-based one).

Forgind replied:
This is for --version, in which case the user does specify an exact version they want. If they just do `dotnet workload update`, it goes through a different path that ultimately picks the highest version.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/FileBasedInstaller.cs:
```diff
@@ -32,6 +32,7 @@ internal class FileBasedInstaller : IInstaller
         private readonly FileBasedInstallationRecordRepository _installationRecordRepository;
         private readonly PackageSourceLocation _packageSourceLocation;
         private readonly RestoreActionConfig _restoreActionConfig;
+        private string workloadSetPath;
```
The convention is for private fields to start with an underscore.  That said, I don't think we will need this field at all if we get rid of the RollBackWorkloadSetInstallation method.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/update/WorkloadUpdateCommand.cs:
```diff
@@ -180,10 +229,13 @@ public void UpdateWorkloads(bool includePreviews = false, DirectoryPath? offline
                         _workloadInstaller.RemoveManifestsFromInstallState(sdkFeatureBand);
                     }
+
+                    _workloadInstaller.AdjustWorkloadSetInInstallState(sdkFeatureBand, workloadSetVersion);
```
It looks like whenever we update to a new workload set, we write the workload set version to the install state (I think this applies to both the update and install commands).

We actually don't normally want to write the workload set version to the install state.  The reason to have a workload set version in the install state is to pin to that version, which should only happen if the `--version` parameter is explicitly specified.  Once the version is pinned, a workload install command should by default not update to a newer version.  A `dotnet workload update --version <version>` command should update the pinned version in the install state, and `dotnet workload update` should remove the pinned version and update to the latest workload set version available.

Forgind replied:
We can talk about this one. I find it very valuable to have the workload set version in the install state file 100% of the time after updating or installing via a workload set. It doesn't have the same 'pinned' function as it does for workload manifests because whereas you can update some workload manifests without updating the pinned workload manifests, the same isn't true for workload sets: there's only one workload set. If you do a `dotnet workload update`, it updates the one and only workload set. If you do `dotnet workload update --version <version>`, it updates the workload set. There isn't really the same kind of scenario where you wouldn't want to update some workload set whereas you would want to update another one if that makes sense.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs:
```diff
@@ -182,6 +185,66 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
+                ModifyWorkloadSet(_newWorkloadSetPath, InstallAction.Uninstall);
+                if (_previousWorkloadSetVersion is not null)
+                {
+                    File.WriteAllText(Path.Combine(_newWorkloadSetPath, "version.txt"), _previousWorkloadSetVersion);
```
What is this version.txt file?  Is this writing to the advertising manifest package?

Forgind replied:
Yes—this was building off of a suggestion from @MiYanni that I rather liked. When we're initially downloading the workload set package, we know what version we ended up downloading, but when it comes time to actually use it, that information is gone, and I couldn't find an easy way to get it back, so at that point, I instead create a version.txt file in the advertising manifest package (which is basically a scratch package, as far as I'm concerned) and parse it later to get the version back. Here, I'm taking advantage of that mechanism to pretend that we'd written the previous workload set version into the version.txt file, which means that when it goes to install a new workload set, it will install the previous one—that is, the one specified in that version file.

Forgind replied:
Make it more explicit on kind of version (packageVersion vs. workloadSetVersion)

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs:
```diff
@@ -182,6 +185,66 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
+        {
+            ReportPendingReboot();
+
+            // Rolling back a manifest update after a successful install is essentially a downgrade, which is blocked so we have to
+            // treat it as a special case and is different from the install failing and rolling that back, though depending where the install
+            // failed, it may have removed the old product already.
+            // Resolve the package ID for the manifest payload package
```
This isn't accurate anymore, as manifests (and workload sets) are always side-by-side, so we won't be upgrading or downgrading them.  https://github.com/dotnet/sdk/pull/38143 simplifies (ie removes) a lot of this logic.

dsplaisted reviewed (commented):
I took a high level look at this and left some feedback about a few things.

I think that this is not yet accounting for the mapping between user-facing workload set versions and the version of the NuGet package for the workload set.  The file-based installer will need to map from the package version to the workload set version and install into the workload set version directory.  The workload manifest updater (or whatever code reads the advertised workload set) will also need to do the same mapping to figure out what the latest workload set version is.

The installers will also need to map the other way, from the workload set version to the package version if they need to download a specific workload set version.

Also, when updating the workload set, there should be an output message that says something like "Updated from workload set 8.0.203 to 8.0.203.1".

Here what we've settled on (so far at least) for how the mapping between these versions should work:

| SDK Version | Workload set version | Workload set Package ID | Workload set package version |
|-------------|----------------------|-------------------------|------------------------------|
| 8.0.200     | 8.0.200              | Microsoft.Net.Workloads.8.0.200 | 8.200.0 |
| 8.0.201     | 8.0.201              | Microsoft.Net.Workloads.8.0.200 | 8.201.0 |
| 8.0.203     | 8.0.203.1            | Microsoft.Net.Workloads.8.0.200 | 8.203.1 |
| 9.0.100-preview.2 | 9.0.100-preview.2  | Microsoft.Net.Workloads.9.0.100-preview.2 | 9.100.0 |
| 8.0.201     | 8.0.201.1-preview    | Microsoft.Net.Workloads.8.0.200 | 8.201.1-preview |
| 8.0.201     | 8.0.201.1-preview.2    | Microsoft.Net.Workloads.8.0.200 | 8.201.1-preview.2 |
| 8.0.201-servicing.23015    | 8.0.201-servicing.23015  | Microsoft.Net.Workloads.8.0.200 | 8.201.0-servicing.23015 |

Forgind commented on src/Cli/dotnet/commands/dotnet-workload/install/IInstaller.cs:
```diff
@@ -13,6 +13,10 @@ internal interface IInstaller : IWorkloadManifestInstaller
     {
         int ExitCode { get; }
 
+        string InstallWorkloadSet(string path);
```
Can you explain why you think that's a better design? From what I see, that would basically mean downloading the nuget package (admittedly potentially from the cache, at least the second time) twice and would require duplicating a lot of logic around calculating manifest updates and executing them. As I have it, it downloads the workload set at the time it would otherwise be looking for advertising manifests. It's important that it installs the workloadset.json file _before_ it tries to calculate manifest updates because it needs that file to calculate manifest updates. And manifest update calculation has to happen before we can actually install the manifests.

Forgind commented on src/Cli/dotnet/commands/dotnet-workload/install/IInstaller.cs:
```diff
@@ -13,6 +13,10 @@ internal interface IInstaller : IWorkloadManifestInstaller
 
+        string InstallWorkloadSet(string path);
+
+        void RollBackWorkloadSetInstallation();
```
I think I implemented this right. Let me know if it looks wrong to you.

Forgind commented:
> I think that this is not yet accounting for the mapping between user-facing workload set versions and the version of the NuGet package for the workload set.

I'll try to get to this tomorrow.

Forgind committed:
PR comments

Forgind commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs:
```diff
@@ -182,6 +182,57 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
+                rollback: () =>
+                {
+                    ModifyWorkloadSet(path, InstallAction.Uninstall);
+                    File.WriteAllText(Path.Combine(path, "version.txt"), previousVersion);
```
Doesn't roll back during gc, so delete these two lines

Forgind commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs:
```diff
@@ -182,6 +182,57 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
             }
         }
 
+        public string InstallWorkloadSet(CliTransaction transaction, string path)
```
For all 'path' references, make it clear (pathToExtractedWorkloadSet) and include comments if possible

Forgind committed:
Address feedback

baronfel commented:
Note that this will need to be retargeted to release/8.0.3xx now/

Forgind commented:
> Note that this will need to be retargeted to release/8.0.3xx now/

I do agree we need to make sure to do this before merging. I've been holding off until I've had a chance to update to the new workload set versioning schema, since we have workload sets for 8.0.2 but not 8.0.3, but I haven't had much of a chance to work the past couple weeks.

Forgind committed:
Convert to and from package version

Forgind committed:
Update version conversion logic

To align with joeloff's version (https://github.com/dotnet/arcade/pull/14526)

Forgind committed:
Add workload set update message

Forgind committed:
Merge branch 'release/8.0.3xx' of https://github.com/dotnet/sdk into sets-in-file-based

Forgind committed:
Delete duplicate method (?)

Forgind committed:
Fix method signature

(and this time, build to make sure that's it)

Forgind committed:
Fix test

Version is not written into the install state by default anymore, so this writes it into a temporary "versionFile" (analogous to packageVersion.txt) and verifies that it's updated to the right workload set version.

Forgind committed:
I'm dumb sometimes

Forgind commented:
Windows_NT Build_Release failing due to CanBuildProjectWithAPackageReferenceWithMultipleAliases.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:
```diff
@@ -72,12 +73,38 @@ private static WorkloadManifestUpdater GetInstance(string userProfileDir)
+                // this updates all the manifests 
+                var manifests = _workloadResolver.GetInstalledManifests();
+                await Task.WhenAll(manifests.Select(manifest => UpdateAdvertisingManifestAsync(manifest, includePreviews, offlineCache))).ConfigureAwait(false);
+                WriteUpdatableWorkloadsFile();
```
This file is used for a message we print out on `dotnet` commands letting people know that a workload update is available.  We should do a similar thing for workload sets.

However, it doesn't need to be in this PR, we can do that as a follow-up.

dsplaisted reviewed (commented):
It looks like this PR is feature-complete for merging, thanks.

Some feedback (which is also covered in PR comments on the code):

- Terminology: We should use "workload version" instead of "workload set" or "workload set version" when possible in user-facing messages.
- Resolver changes / workload set filenames: I sent an email about this.  I think we want to revert most of the SdkDirectoryWorkloadManifestProvider changes here and re-generate the workload set packages with different .json filenames.
- Pinning: We shouldn't write the workload set version to the install state file unless `--version` was specified (see my PR code comment for more details)
- Transactions: The `CliTransaction` class isn't being used correctly, which I think means it wouldn't actually run the rollback actions.

Follow-up items (ie things to do in subsequent PRs):

- Offline cache: We support a `--download-to-cache` parameter to download all required packages locally, and a `--from-cache` parameter to use that to do updates / installs without network access.  It looks like those aren't supported for workload sets here.  Those parameters are hidden and a lot lower priority, so we can add support for them as a follow-up item (if at all).  FYI @marcpopMSFT @baronfel 
- Add support for `dotnet workload restore` with workload set version specified in global.json, as well as root tracking, garbage collection, etc.
- Check for available workload set updates and notify user about them (see WriteUpdatableWorkloadsFile)
- Tests: I'd like to see a lot more workload set tests.  I think the VM testing infrastructure I'm working on will be great for this.  I also have some ideas for file-based end-to-end workload tests, but probably that's a lot longer term.
- ManifestVersionUpdate class: This isn't directly related to this PR or to workload sets, but I noticed it when reviewing the code.  Do we need to represent the existing/old manifest version in `ManifestVersionUpdate` anymore?  Manifests are now side-by-side so it's probably not needed.

We had also previously discussed how the APIs here are passing around information such as the workload set version via a path to the package in the advertising manifests.  I still don't think that this is the best design.  If we are worried about downloading packages twice, I took a look and for MSI-based installs we don't actually do so because the MSI installer manages its own cache of downloaded packages.  If the double download is actually an issue we could do the same type of caching for the file-based installer.

However, I don't want to block on changing this design, so we can check it in as is and possibly refactor it later.

Forgind commented on src/Cli/dotnet/commands/dotnet-workload/update/WorkloadUpdateCommand.cs:
```diff
@@ -175,6 +224,8 @@ public void UpdateWorkloads(bool includePreviews = false, DirectoryPath? offline
                         _workloadInstaller.RemoveManifestsFromInstallState(sdkFeatureBand);
                     }
 
+                    _workloadInstaller.AdjustWorkloadSetInInstallState(sdkFeatureBand, string.IsNullOrWhiteSpace(_workloadSetVersion) ? null : workloadSetVersion);
```
That actually is not what this does. If the version isn't specified explicitly (i.e., _workloadSetVersion is null or white space), then it writes null into the install state for the workload set version, and since it doesn't serialize nulls, that means it _removes_ the version (if present) from the install state.

dsplaisted replied:
I traced the code and it looked like `_workloadSetVersion` would be set here even if it wasn't specified on the command line, because there was code in one of the calling methods that read it from the advertised workload set and passed it along.

Forgind replied:
This is the total set of writes to _workloadSetVersion:
<img width="531" alt="image" src="https://github.com/dotnet/sdk/assets/12969783/db2df5f0-7ecd-4fbf-ab60-35a3632df15b">

Two of those are clearly from the parseResult, i.e., from the command line. The third is here with this comment:
<img width="750" alt="image" src="https://github.com/dotnet/sdk/assets/12969783/b7c8a14e-dce4-4fe5-a72d-699e63e9cc0e">

In other words, if it's already pinned, we treat it as if it's pinned so that it stays in the install state file. If you run a vanilla `dotnet workload update`, it isn't set.

dsplaisted replied:
I think maybe I got confused here by the difference between `_workloadSetVersion` and `workloadSetVersion`.  I commented on the same code in `WorkloadInstallCommand`, I think it would be easier to understand if this line was consistent about using `_workloadSetVersion`, and it would be good to have a comment explaining why it does so.

Forgind commented on src/Cli/dotnet/commands/dotnet-workload/install/FileBasedInstaller.cs:
```diff
@@ -85,6 +85,27 @@ IEnumerable<PackInfo> GetPacksInWorkloads(IEnumerable<WorkloadId> workloadIds)
+                {
+                    Directory.CreateDirectory(Path.GetDirectoryName(workloadSetPath));
+
+                    File.Copy(Path.Combine(advertisingPackagePath, "workloadset.json"), workloadSetPath, overwrite: true);
```
The previous design was that it was any file that ended in .workloadset.json, but my understanding is that that's no longer true; now there's just one workloadset.json for any given version. Did that change again @marcpopMSFT?

Forgind commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:
```diff
@@ -157,13 +157,18 @@ public void RefreshWorkloadManifests()
                 return _workloadSet?.Version!;
             }
 
+            if (InstallStateContents.FromPath(Path.Combine(WorkloadInstallType.GetInstallStateFolder(_sdkVersionBand, _sdkRootPath), "default.json")).UseWorkloadSets == true)
```
Isn't it lying if the user expects a workload set version, and we give them a hash of a bunch of random manifests? It's a unique version, but it isn't a workload set version, so if they try to download it later, they'll get an error.

Forgind replied:
Note that I only have it returning null here if the user requests workload sets, but we haven't installed one yet.

Forgind commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.GlobalJsonReader.cs:
```diff
@@ -61,7 +61,7 @@ static class GlobalJsonReader
                                         case JsonTokenType.PropertyName:
                                             var sdkPropName = reader.GetString();
-                                            if (string.Equals("workloadVersion", sdkPropName, StringComparison.OrdinalIgnoreCase))
+                                            if (string.Equals("workloadSetVersion", sdkPropName, StringComparison.OrdinalIgnoreCase))
```
My understanding was that we didn't have workload set version support for global.json yet, so I didn't think of it as a breaking change yet and preferred the more descriptive name. Both because it sounds like the current consensus is to use "workload version" instead of "workload set version" and because of what you said about updating VS, I don't think this is a good idea anymore, so I'll revert it.

Forgind committed:
Most PR comments

Forgind committed:
Fix transactions and other little things

dsplaisted commented on src/Cli/Microsoft.DotNet.Cli.Utils/PathUtility.cs:
```diff
@@ -97,6 +97,30 @@ public static bool TryDeleteDirectory(string directoryPath)
+        /// and continues to its parent until it fails. Returns whether it succeeded
+        /// in deleting the file it was intended to delete.
+        /// </summary>
+        public static bool DeleteFileAndEmptyParents(string path)
```
I think under the dotnet/metadata directory there might be directories that we don't want to delete even if they are empty.  It's not a big deal though.

Forgind committed:
Permit multiple *.workloadset.json files

Forgind committed:
Fix transaction usage + update some tests

Forgind committed:
Correct my test

Forgind committed:
Merge branch 'release/8.0.3xx' into sets-in-file-based

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -236,6 +267,8 @@ private void WriteSDKInstallRecordsForVSWorkloads()
                         installer.SaveInstallStateManifestVersions(sdkFeatureBand, GetInstallStateContents(manifestsToUpdate));
                     }
 
+                    installer.AdjustWorkloadSetInInstallState(sdkFeatureBand, string.IsNullOrWhiteSpace(_workloadSetVersion) ? null : workloadSetVersion);
```
It seems a bit confusing that this uses two different values: `_workloadSetVersion` and `workloadSetVersion`.  Is there any case where `_workloadSetVersion` would be set, but `workloadVersion` would have a different value?  I think not, so I think this can use `_workloadSetVersion` in both cases.  If not, then this should definitely have a comment explaining the difference.

Probably it would be good to have a comment anyway explaining why this uses `_workloadSetVersion` and not `workloadSetVersion`.

Forgind replied:
The only way I think _workloadSetVersion could be set, and it would be different from workloadSetVersion is if the user requests version x, and we reach out to NuGet and ask for x, and it returns version y instead. I don't think that's possible today—I think if NuGet can't find the right version, it would error rather than return the wrong version—so I think you're right that I can switch to just using _workloadSetVersion and delete workloadSetVersion. I should be able to do something similar in WorkloadUpdateCommand.

dsplaisted replied:
I don't think you want to delete `workloadSetVersion` entirely, just don't use it on this line.  If there's no pinned workload set version and no version was specified on the command line, but the install state is opted into workload sets, then `_workloadSetVersion` would be null, but `workloadSetVersion` would be the latest workload set version that was found, and should be installed / used.

Forgind replied:
Right. We'd still install that version, but we don't actually need to keep track of it in Workload(Update/Install)Command.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:
```diff
@@ -261,64 +303,57 @@ public async Task<IEnumerable<WorkloadDownload>> GetManifestPackageDownloadsAsyn
+                    // If an offline cache is present, use that. Otherwise, try to acquire the package online.
+                    packagePath = offlineCache != null ?
+                        Directory.GetFiles(offlineCache.Value.Value)
+                            .Where(path => path.EndsWith(".nupkg") && Path.GetFileName(path).StartsWith(manifestPackageId.ToString(), StringComparison.OrdinalIgnoreCase))
+                            .Max() :
```
This looks like the correct logic if `packageVersion` is not specified.  If it is specified, I think this logic needs to take the version into account to return the package version specified instead of the latest version in the cache.

Forgind replied:
That was how it was before...I'll admit I thought that was a little odd but figured it must not matter too much since we hadn't fixed it for other manifests? But I can change the logic for this.

dsplaisted replied:
I think the difference is that previously it would always look for the latest version of a package, it didn't have the capability of specifying a specific version.

Forgind replied:
Mmm...I think you're right. I'd thought there was something that handled versions previously, but I don't know what I was thinking of.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:
```diff
@@ -419,7 +470,41 @@ private async Task<bool> NewerManifestPackageExists(ManifestId manifest)
         }
 
-        private IEnumerable<(ManifestId Id, ManifestVersionWithBand ManifestWithBand)> ParseRollbackDefinitionFile(string rollbackDefinitionFilePath)
+        public IEnumerable<ManifestVersionUpdate> ParseRollbackDefinitionFiles(IEnumerable<string> rollbackFilePaths)
```
`SdkDirectoryWorkloadManifestProvider.GetAvailableWorkloadSets` has logic that does the same thing as this method.  Could this be factored so there's only one implementation?

Forgind replied:
They don't look like they do the same thing. That method finds all installed workload sets and returns the ones that match this feature band. This takes in paths to workload sets and parses them into a set of manifest version updates. It's basically "what happens after you call SdkDirectoryWorkloadManifestProvider.GetAvailableWorkloadSets"

dsplaisted replied:
The methods don't do the same thing, but inside the loop in `GetAvailableWorkloadSets` it seemed like it was doing the same thing (aggregating multiple json files into a single workload set).

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:
```diff
@@ -419,7 +470,41 @@ private async Task<bool> NewerManifestPackageExists(ManifestId manifest)
+            fullSet = fullSet.DistinctBy<(ManifestId, ManifestVersionWithBand), ManifestId>(update => update.Item1).ToList();
+            if (size != fullSet.Count)
+            {
+                throw new Exception();
```
We shouldn't throw an exception with no details here.  It looks like this is probably supposed to catch the case where the same manifest ID appears in multiple json files in the workload set package.  If so, ideally the message would include information about which ID was duplicated.  However, since in practice we don't expect multiple json files in the package, it's not a big deal.  But we still shouldn't throw an Exception without a message.

Forgind replied:
Good catch; that was supposed to have a real exception, but then I forgot to add that part 🙂

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/update/WorkloadUpdateCommand.cs:
```diff
@@ -118,29 +137,56 @@ public override int Execute()
+            {
+                // Rollback files are only for loose manifests. Update the mode to be loose manifests.
+                Reporter.WriteLine(LocalizableStrings.UpdateFromRollbackSwitchesModeToLooseManifests);
+                _workloadInstaller.UpdateInstallMode(_sdkFeatureBand, false);
```
This should probably be run inside the transaction, and rolled back if the update fails.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/update/WorkloadUpdateCommand.cs:
```diff
@@ -232,5 +269,21 @@ private IEnumerable<WorkloadId> GetUpdatableWorkloads()
             return workloads;
         }
+
+        private void RunInNewTransaction(Action<ITransactionContext> a)
```
Is this the exact same code as in `WorkloadInstallCommand`?  If so, should this be moved up to the base class?

Forgind replied:
I actually had the exact same thought. The only issue is that the strings are separate, so this is an annoyingly big change. I do think it would be cleaner to do it that way, though...

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/InstallStateContents.cs:
```diff
@@ -12,17 +14,21 @@ internal class InstallStateContents
+        public Dictionary<string, string>? Manifests { get; set; }
+
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+        public string? WorkloadSetVersion { get; set; }
```
I think I mentioned this when we were discussing the PR, but I may have forgotten to include it in the PR comments: The property name in the global.json file should be `workloadVersion` (though it looks like the tests weren't consistent with the casing, so probably we should be case-insensitive).

So this property should be named `WorkloadVersion`, or I think you could also use a Json attribute to change the name used when reading / writing json.

dsplaisted commented on src/Tests/Microsoft.NET.Sdk.WorkloadManifestReader.Tests/SdkDirectoryWorkloadManifestProviderTests.cs:
```diff
@@ -605,7 +605,7 @@ public void ItUsesWorkloadSetFromInstallState()
                 """
                 {
-                    "workloadVersion": "8.0.201"
+                    "workloadSetVersion": "8.0.201"
```
As mentioned in `InstallStateContents`, the JSON property should be `workloadVersion`, so these changes in the tests should be reverted.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:
```diff
@@ -134,6 +161,22 @@ public void DeleteUpdatableWorkloadsFile()
+            return $"{nugetVersion.Major}.{nugetVersion.Patch}.{patch}{release}";
+        }
+
+        public static string WorkloadSetPackageVersionToWorkloadSetVersion(SdkFeatureBand sdkFeatureBand, string packageVersion)
```
It looks like this wasn't addressed yet, it would be good to sync the implementations with arcade, especially since there are tests covering arcade's version of this.

dsplaisted reviewed (approved):
This is quite close.  I had a few comments, the most important to change before merging is the name of the `workloadVersion` property in global.json.

Forgind committed:
More feedback
